### PR TITLE
nixos/nvidia : added nvidia-persistenced

### LINKED ISSUE #75099: Add an option for NVIDA GPU headless mode, i.e. nvidia-persistenced
As described here https://github.com/NixOS/nixpkgs/issues/73804 it would be useful to add an option in `hardware.nvidia` to add a systemd entry which forces GPU's to always be online even in headless systems.

Here is what the extracted systemd service would look like (taken from https://github.com/NixOS/nixpkgs/issues/73804#issuecomment-562469900)

```
systemd.services."nvidia-persistenced" = {
  description = "NVIDIA Persistence Daemon";
  wantedBy = [ "multi-user.target" ];
  serviceConfig = {
    Type = "forking";
    Restart = "always";
    PIDFile = "/var/run/nvidia-persistenced/nvidia-persistenced.pid";
    ExecStart = "${pkgs.linuxPackages.nvidia_x11.persistenced}/bin/nvidia-persistenced --verbose";
    ExecStopPost = "${pkgs.coreutils}/bin/rm -rf /var/run/nvidia-persistenced";
  };
};
```

### PR DESCRIPTION
**Pull request by Rakesh4G**
###### Motivation for this change
Fixes: https://github.com/NixOS/nixpkgs/issues/75099

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @CMCDragonkai @mdedetrich

### CODE REVIEW

**Comment by mdedetrich** [comment]
I believe this should be optional with default being turned off, reason being its when you have NVidia GPU's that are used for machine learning so you want the Nvidia GPU's to always be online even though your display is being run through a completely different GPU (i.e. Intel integrated GPU).

@CMCDragonkai Can you confirm this?

**Comment by CMCDragonkai** [comment]
Yes by default it should be off.

**Comment by Rakesh4G** [comment]
OK.

**Comment by CMCDragonkai** [comment]
This requires a module option and a `mkIf` that checks the enable option before writing the config.

**Comment by Rakesh4G** [comment]
> This requires a module option and a `mkIf` that checks the enable option before writing the config.

Done changes to make this optional and default to `false`.
@CMCDragonkai @mdedetrich

**Comment by mdedetrich** [comment]
Looks fine to me, @CMCDragonkai ?

**Comment by coreyoconnor** [comment]
Works for me: nvidia-smi consistently reports both GPUs available

~~~
Tue Jan 21 17:46:14 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.44       Driver Version: 440.44       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K10.G1.8GB    On   | 00000000:05:00.0 Off |                  Off |
| N/A   29C    P8    17W / 117W |      0MiB /  4037MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K10.G1.8GB    On   | 00000000:06:00.0 Off |                  Off |
| N/A   37C    P8    17W / 117W |      0MiB /  4037MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

~~~

**Comment by CMCDragonkai** [comment]
Your machine is different. This is needed headless servers with embedded intel gpu and 4 or more gpus.

On 22 January 2020 12:47:07 GMT+11:00, coreyoconnor <notifications@github.com> wrote:
>Works for me: nvidia-smi consistently reports both GPUs available
>
>~~~
>Tue Jan 21 17:46:14 2020       
>+-----------------------------------------------------------------------------+
>| NVIDIA-SMI 440.44       Driver Version: 440.44       CUDA Version:
>10.2     |
>|-------------------------------+----------------------+----------------------+
>| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile
>Uncorr. ECC |
>| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util 
>Compute M. |
>|===============================+======================+======================|
>|   0  Tesla K10.G1.8GB    On   | 00000000:05:00.0 Off |               
>  Off |
>| N/A   29C    P8    17W / 117W |      0MiB /  4037MiB |      0%     
>Default |
>+-------------------------------+----------------------+----------------------+
>|   1  Tesla K10.G1.8GB    On   | 00000000:06:00.0 Off |               
>  Off |
>| N/A   37C    P8    17W / 117W |      0MiB /  4037MiB |      0%     
>Default |
>+-------------------------------+----------------------+----------------------+
>                                                                       
>+-----------------------------------------------------------------------------+
>| Processes:                                                       GPU
>Memory |
>|  GPU       PID   Type   Process name                            
>Usage      |
>|=============================================================================|
>|  No running processes found                                          
>      |
>+-----------------------------------------------------------------------------+
>
>~~~
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/77054#issuecomment-576970220

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by coreyoconnor** [comment]
> Your machine is different. This is needed headless servers with embedded intel gpu and 4 or more gpus.
> […](#)
> On 22 January 2020 12:47:07 GMT+11:00, coreyoconnor ***@***.***> wrote: Works for me: nvidia-smi consistently reports both GPUs available ~~~ Tue Jan 21 17:46:14 2020 +-----------------------------------------------------------------------------+ | NVIDIA-SMI 440.44 Driver Version: 440.44 CUDA Version: 10.2 | |-------------------------------+----------------------+----------------------+ | GPU Name Persistence-M| Bus-Id Disp.A | Volatile Uncorr. ECC | | Fan Temp Perf Pwr:Usage/Cap| Memory-Usage | GPU-Util Compute M. | |===============================+======================+======================| | 0 Tesla K10.G1.8GB On | 00000000:05:00.0 Off | Off | | N/A 29C P8 17W / 117W | 0MiB / 4037MiB | 0% Default | +-------------------------------+----------------------+----------------------+ | 1 Tesla K10.G1.8GB On | 00000000:06:00.0 Off | Off | | N/A 37C P8 17W / 117W | 0MiB / 4037MiB | 0% Default | +-------------------------------+----------------------+----------------------+ +-----------------------------------------------------------------------------+ | Processes: GPU Memory | | GPU PID Type Process name Usage | |=============================================================================| | No running processes found | +-----------------------------------------------------------------------------+ ~~~ -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#77054 (comment)](https://github.com/NixOS/nixpkgs/pull/77054#issuecomment-576970220)
> -- Sent from my Android device with K-9 Mail. Please excuse my brevity.

I think you are confused. This is useful on any system using a nvidia GPU in a headless mode. As that is my system *and* this definitely resolves issues the requirements of "embedded intel gpu and 4 or more gpus" is incorrect. That would also be inconsistent with the documentation of nvidia-persistenced.

**Comment by CMCDragonkai** [comment]
Great, then this works for any headless system then.

**Comment by CMCDragonkai** [comment]
This is ready to be merged. Anybody?

**Comment by CMCDragonkai** [comment]
@Rakesh4G Can you clean up the commits, and squash them into 1, and abide by the commit standard.

**Comment by Rakesh4G** [comment]
Hi @CMCDragonkai , I have cleaned up the commits, and squash them into 1. Thanks.

File: nixos/modules/hardware/video/nvidia.nix [hunk 1]
```diff
@@ -186,6 +195,18 @@ in
...
         "L+ /run/nvidia-docker/extras/bin/nvidia-persistenced - - - - ${nvidia_x11.persistenced}/origBin/nvidia-persistenced";
 
+    systemd.services."nvidia-persistenced" = mkIf nvidiaPersistencedEnabled {
+      description = "NVIDIA Persistence Daemon";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "forking";
+        Restart = "always";
+        PIDFile = "/var/run/nvidia-persistenced/nvidia-persistenced.pid";
+        ExecStart = "${pkgs.linuxPackages.nvidia_x11.persistenced}/bin/nvidia-persistenced --verbose";
```
Comment (CMCDragonkai) [inline nixos/modules/hardware/video/nvidia.nix]: Change `pkgs.linuxPackages.nvidia_x11.persistenced` to `nvidia_x11.persistenced`.

**Comment by CMCDragonkai** [comment]
You have to resolve the conflicts in the MR right now.

But it's unfortunate that this MR has been left without any one attempting to push it in.

File: nixos/modules/hardware/video/nvidia.nix [hunk 2]
```diff
@@ -106,6 +107,14 @@ in
         shows the Intel GPU at "00:02.0", set this option to "PCI:0:2:0".
       '';
     };
+
+    hardware.nvidia.nvidiaPersistenced = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        Update for NVIDA GPU headless mode, i.e. nvidia-persistenced.
```
[the remark this answers is not in the document]
Comment (Rakesh4G) [reply nixos/modules/hardware/video/nvidia.nix]: ok

File: nixos/modules/hardware/video/nvidia.nix (hunk 1 above)
Comment (Rakesh4G) [reply nixos/modules/hardware/video/nvidia.nix]: ok.

**Comment by CMCDragonkai** [comment]
This is ready to be merged.

Anyone?

**Review by CMCDragonkai — APPROVED** [review]
Approved.

@matthewbauer 

@worldofpeace

**Comment by mdedetrich** [comment]
I have no permission to merge, I think you will have to ping a maintainer.

**Comment by CMCDragonkai** [comment]
Who's the maintainer for this module?

**Comment by mdedetrich** [comment]
Judging by https://github.com/NixOS/nixpkgs/pull/66601 @romildo seems to merge NVidia related PR's

**Pull request merged**